src/dstack/_internal/server/background/tasks/process_instances.py

```diff
@@ -99,54 +100,249 @@ async def process_instances() -> None:
                             InstanceStatus.TERMINATING,
                         ]
                     ),
-                    InstanceModel.id.not_in(PROCESSING_INSTANCES_IDS),
+                    InstanceModel.id.not_in(lockset),
                 )
+                .options(lazyload(InstanceModel.job))
+                .order_by(InstanceModel.last_processed_at.asc())
+                .limit(1)
+                .with_for_update(skip_locked=True)
             )
-            instances = res.all()
```

Comment:
Before the PR, instances were processed concurrently with `asyncio.as_completed` which cannot be done when the processing happens in the scope of one transaction (it's not safe to share an sqlalchemy session between coroutines). Processing multiple instances both concurrently and sequentially is also problematic because instance processing can take time (e.g. remote instance waiting) which would lead to blocking processing. The current approach tolerates some instances being slow to process. It's consistent with other processing tasks.

If the processing throughput/latency becomes a problem, we could just decrease processing interval / increase max_instances.